packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,339 @@
+  );
+};
+
+/**
+ * Checks if the timestamp is within a specified range.
+ * @param timestamp {number} - The timestamp to check
+ * @param startTimestamp {number} - The start timestamp
+ * @param endTimestamp {number} - The end timestamp
+ * @returns {boolean}
+ */
+const isTimestampInRange = (
+  timestamp: number,
+  startTimestamp: number,
+  endTimestamp: number
+): boolean => timestamp >= startTimestamp && timestamp <= endTimestamp;
+
+/**
+ * Inserts the session in the timeline at the right location.
+ * @param currentSession {UserSession} - The current session
+ * @param ticketCreationDate {number} - The ticket creation date
+ * @param timeline {UserActionTimeline} - The timeline
+ * @returns {void}
+ */
+const insertSessionInTimeline = (
+  session: UserSession,
+  ticketCreationDate: number,
+  timeline: UserActionTimeline
+) => {
+  if (ticketCreationDate < session.start) {
+    timeline.followingSessions.push(session);
+  } else {
+    timeline.precedingSessions.push(session);
+  }
+};
+
+/**
+ * Divides actions into sessions and organizes them in the timeline based on the ticket creation date. This function does the following:
+ * 1) Iterates over the actions and groups them into sessions based on the session inactivity threshold.
+ * 2) Inserts the case creation action in the current session.
+ * 3) Inserts the session in the timeline at the right location.
+ * 4) Returns the timeline with the current session, 2 preceding sessions and 2 following sessions.
+ * @param actions {UserAction[]} - The actions to split
+ * @param ticketCreationDate {string} - The ticket creation date
+ * @returns {UserActionTimeline} - The timeline of user actions (current session and 2 preceding and following sessions)
+ */
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: number
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isActionWithinSessionThreshold(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+    // We check if the case creation is within the current session or within the inactivity threshold after the end of the session.
+    const isCaseCreationIsPartOfCurrentSession = isTimestampInRange(
+      ticketCreationDate,
+      currentSession.start,
+      currentSession.end + SESSION_INACTIVITY_THRESHOLD_IN_MS
+    );
+
+    if (isCaseCreationIsPartOfCurrentSession) {
+      let caseCreationActionInserted = false;
+      const caseCreationAction: UserAction = {
+        actionType: UserActionType.TICKET_CREATION,
+        timestamp: ticketCreationDate,
+        eventData: {},
+      };
```

Comment:
There is a problem with this,

If for example the received actions are the following:
```
[{value: 'action one', time:1},{value: 'action two', time:3}]
``` 

and the case creation timestamp corresponds to `2`, the block of code that checks if the case is part of the session will never be reached because of the return in line `198`